### PR TITLE
CLN: Split and fixturized test_fillna in tests/base/test_ops.py

### PR DESCRIPTION
**Pull request by SaturnFromTitan**
part of #23877
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: pandas/tests/base/test_ops.py [hunk 1]
```diff
@@ -719,58 +719,63 @@ def test_drop_duplicates_series_vs_dataframe(self):
...
+        else:
+            tm.assert_series_equal(obj, result)
 
-                if not allow_na_ops(o):
-                    continue
+        # check shallow_copied
+        assert obj is not result
 
-                if needs_i8_conversion(o):
+    @pytest.mark.parametrize("null_obj", [np.nan, None])
```
Comment (SaturnFromTitan) [inline pandas/tests/base/test_ops.py]: Using `nulls_fixture` should be taken care of in #32437

File: pandas/tests/base/test_ops.py [hunk 2]
```diff
@@ -719,58 +719,63 @@ def test_drop_duplicates_series_vs_dataframe(self):
...
-            if isinstance(o, Index):
-                tm.assert_index_equal(o, result)
-            else:
-                tm.assert_series_equal(o, result)
-            # check shallow_copied
-            assert o is not result
+        obj = index_or_series_obj
+        if len(obj) == 0:
+            pytest.skip("Test doesn't make sense on empty data")
+        elif isinstance(obj, pd.MultiIndex):
```
Comment (MomIsBestFriend) [inline pandas/tests/base/test_ops.py]: Can you use ```ABCMultiIndex``` from ```pandas.core.dtypes.generic```?

File: pandas/tests/base/test_ops.py [hunk 3]
```diff
@@ -719,58 +719,63 @@ def test_drop_duplicates_series_vs_dataframe(self):
...
+            pytest.skip("MultiIndex doesn't support isna")
 
-        for null_obj in [np.nan, None]:
-            for orig in self.objs:
-                o = orig.copy()
-                klass = type(o)
+        # values will not be changed
+        result = obj.fillna(obj.values[0])
+        if isinstance(obj, Index):
```
Comment (MomIsBestFriend) [inline pandas/tests/base/test_ops.py]: Can you use ```ABCIndex``` from ```pandas.core.dtypes.generic```?

File: pandas/tests/base/test_ops.py [hunk 4]
```diff
@@ -719,58 +719,63 @@ def test_drop_duplicates_series_vs_dataframe(self):
...
-                    values[0:2] = pd.NaT
-                else:
-                    values = o.values.copy()
-                    fill_value = o.values[0]
-                    values[0:2] = null_obj
+        if not allow_na_ops(obj):
+            pytest.skip(f"{klass} doesn't allow for NA operations")
+        elif len(obj) < 1:
+            pytest.skip("Test doesn't make sense on empty data")
+        elif isinstance(obj, pd.MultiIndex):
```
Comment (MomIsBestFriend) [inline pandas/tests/base/test_ops.py]: Can you use ```ABCMultiIndex``` from ```pandas.core.dtypes.generic```?

File: pandas/tests/base/test_ops.py [hunk 5]
```diff
@@ -719,58 +719,63 @@ def test_drop_duplicates_series_vs_dataframe(self):
...
 
-                expected = klass(expected, dtype=orig.dtype)
-                o = klass(values)
+        expected = klass(expected)
+        obj = klass(values)
 
-                # check values has the same dtype as the original
-                assert o.dtype == orig.dtype
+        result = obj.fillna(fill_value)
+        if isinstance(obj, Index):
```
Comment (MomIsBestFriend) [inline pandas/tests/base/test_ops.py]: Can you use ```ABCIndex``` from ```pandas.core.dtypes.generic```?

**Review by ShaharNaveh — CHANGES_REQUESTED** [review]
few nits related to #27353

File: pandas/tests/base/test_ops.py (hunk 2 above)
Comment (SaturnFromTitan) [reply pandas/tests/base/test_ops.py]: Sure, done. There are more cases like that in the file, but I'd rather change these in a follow-up PR

Comment (MomIsBestFriend) [reply pandas/tests/base/test_ops.py]: @SaturnFromTitan That would be great!

**Comment by SaturnFromTitan** [comment]
@MomIsBestFriend Seems like that didn't work for the `isinstance(obj, ABCIndex)` cases because...
```
isinstance(pd.Float64Index([]), ABCIndex) is False
isinstance(pd.IntervalIndex([]), ABCIndex) is False
isinstance(pd.CategoricalIndex([]), ABCIndex) is False
```

I'll revert it to a simple `isinstance(obj, Index)`

File: pandas/tests/base/test_ops.py [hunk 6]
```diff
@@ -719,58 +720,63 @@ def test_drop_duplicates_series_vs_dataframe(self):
...
-            result = o.fillna(o.astype(object).values[0])
-            if isinstance(o, Index):
-                tm.assert_index_equal(o, result)
-            else:
-                tm.assert_series_equal(o, result)
-            # check shallow_copied
-            assert o is not result
+        obj = index_or_series_obj
+        if len(obj) == 0:
+            pytest.skip("Test doesn't make sense on empty data")
```
Comment (WillAyd) [inline pandas/tests/base/test_ops.py]: This should still work no? Maybe just need to pick a different fill value

File: pandas/tests/base/test_ops.py [hunk 7]
```diff
@@ -719,58 +720,63 @@ def test_drop_duplicates_series_vs_dataframe(self):
...
+        elif isinstance(obj, ABCMultiIndex):
+            pytest.skip("MultiIndex doesn't support isna")
```
Comment (WillAyd) [inline pandas/tests/base/test_ops.py]: Is this an open issue?

File: pandas/tests/base/test_ops.py [hunk 8]
```diff
@@ -719,58 +720,63 @@ def test_drop_duplicates_series_vs_dataframe(self):
...
-                else:
-                    values = o.values.copy()
-                    fill_value = o.values[0]
-                    values[0:2] = null_obj
+        if not allow_na_ops(obj):
+            pytest.skip(f"{klass} doesn't allow for NA operations")
+        elif len(obj) < 1:
+            pytest.skip("Test doesn't make sense on empty data")
+        elif isinstance(obj, ABCMultiIndex):
+            pytest.skip(f"MultiIndex can't hold '{null_obj}'")
```
Comment (WillAyd) [inline pandas/tests/base/test_ops.py]: Another open issue?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/base/test_ops.py (hunk 6 above)
Comment (SaturnFromTitan) [reply pandas/tests/base/test_ops.py]: Fair point. I actually discovered a bug by including this, thanks! I added a reference to the github issue as a comment

File: pandas/tests/base/test_ops.py (hunk 8 above)
Comment (SaturnFromTitan) [reply pandas/tests/base/test_ops.py]: Isn't this by design? A MultiIndex must hold values on multiple levels, so its elements are tuples. Therefore a `MultiIndex` can't just hold a regular `np.nan` or `None` as a value. Maybe a tuple of nans might make sense, but I don't think that would be recognised as "na" by fillna.

File: pandas/tests/base/test_ops.py (hunk 7 above)
Comment (SaturnFromTitan) [reply pandas/tests/base/test_ops.py]: See my comment below, this is related to `MultiIndex` not being able to hold "pure" NA values

**Comment by jreback** [comment]
thanks @SaturnFromTitan 

ideally if you can add the nulls fixture would be great :->

**Pull request merged**